doc/manual/src/protocols/signature.md

```diff
@@ -0,0 +1,49 @@
+# Signature
+
+In Nix, signatures are used to trust store paths using asymmetric cryptography,
+in this instance: Curve25519-based cryptography.
+
+## Store path signature format
+
+Store path signatures are ED25519 signatures, there are two types of signatures:
+
+- `ValidPathInfo` will assemble a fingerprint in the form of
+  ```
+  1;<store path>;<NAR hash in base32 including its own type>;<NAR size in bytes>;<NAR references separated by ,>
+  ```
+  and sign this information with the ED25519 private key.
+
+- `Realisation` will assemble a fingerprint in the form of a JSON string:
+  ```
+  { id, outPath, dependentRealisations }
+  ```
+  and sign this information with the ED25519 private key.
+
+# Remote signature protocol
+
+The remote signature protocol is a mechanism to offload the signature of your store paths to another machine that can possess the secret key material in a secure way.
+
+In this setup, Nix will contact a remote signing URL that you specified and ask to sign fingerprints over the wire.
+
+The protocol expects a UNIX domain socket to force you to handle proper authentication and authorization.
+`socat` is a great tool to manipulate all sorts of sockets.
+
+## Semantics of the APIs
+
+- `POST /sign`:
+
+  Expects a fingerprint as input and will return the signature associated to that fingerprint.
+
+- `POST /sign-store-path`:
+
+  Expects a store path as a parameter and will attempt to sign that specific store path which is expected to be present on the signer's machine and return the signature in the response.
```

Comment:
This doesn't exist at the moment, but I wonder whether it should. Yes, I talked about moving things into `libutil` before, but my experience with hardware parse-display-and-sign devices make me wonder if we should always send over a "nice, JSON" form of what is to be signed, and the other end is responsible for concatenating the actual signature data.

If nothing else, forcing the remote side to munge the bytes is a good defense against lazy blind-singing implementations.

Reply:
Well, it "exists" in that I implemented it in my service-side of this PR (https://github.com/cole-h/nixos-cache-signing-server), heh.

I do think sending the bits necessary for signing as a well-formed JSON document is a good idea, and will make it harder to misuse.

(The only reason `/sign-store-path` exists is because I initially implemented it in terms of shelling out to `nix`, because at the time I couldn't figure out how to do the signing as Nix does it. This has been since rectified, so honestly I wouldn't be opposed to removing it altogether, unless there's a reason to keep it around?)

Reply:
As said privately to @Ericson2314, I don't believe this sign store path API is good for us yet because we are looking at a high frequency signature system, e.g. hydra.nixos.org and there's no distributed filesystems that will make those stores paths available to the signer in time so they can sign them.

For technical reasons, we need a blind signer that is sufficiently resistant to misuse to avoid disasters for random people and ourselves.

Sending a well-formed JSON document looks good to me, but how is it harder to misuse? The only way to make it better is to introduce authnz or auth inside the server and the protocol, this was obviously out of scope because we cannot represent all the usecases in such a simple PR.

Once this lands, my opinion is that we should focus our efforts on understanding how to move the signing to the remote server using PKCS#11 API (which is almost ready with `cryptoki` and `rust-cryptoki` thanks to @baloo) and then think from the HSM perspective. There are many missing pieces and knobs around this system of signature: no revocation, no "timestamping", no certification paths, etc. This is understandable why but all those years have passed and those missing features are limiting our operational range in regards to the cache security.